etcdserver/verify/verify.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2021 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package verify
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"go.etcd.io/etcd/mvcc"
+	"go.etcd.io/etcd/mvcc/backend"
+	"go.etcd.io/etcd/pkg/fileutil"
+	"go.etcd.io/etcd/version"
+	"go.uber.org/zap"
+)
+
+const ENV_VERIFY = "ETCD_VERIFY"
+const ENV_VERIFY_ALL_VALUE = "all"
+
+type Config struct {
+	// DataDir is a root directory where the data being verified are stored.
+	DataDir string
+
+	Logger *zap.Logger
+}
+
+// Verify performs consistency checks of given etcd data-directory.
+// The errors are reported as the returned error, but for some situations
+// the function can also panic.
+// The function is expected to work on not-in-use data model, i.e.
+// no file-locks should be taken. Verify does not modified the data.
+func Verify(cfg Config) error {
+	lg := cfg.Logger
+	if lg == nil {
+		lg = zap.NewNop()
+	}
+
+	if !fileutil.Exist(toBackendFileName(cfg.DataDir)) {
+		lg.Info("verification skipped due to non exist db file")
+		return nil
+	}
+
+	var err error
+	lg.Info("verification of persisted state", zap.String("data-dir", cfg.DataDir))
+	defer func() {
+		if err != nil {
+			lg.Error("verification of persisted state failed",
+				zap.String("data-dir", cfg.DataDir),
+				zap.Error(err))
+		} else if r := recover(); r != nil {
+			lg.Error("verification of persisted state failed",
+				zap.String("data-dir", cfg.DataDir))
+			panic(r)
+		} else {
+			lg.Info("verification of persisted state successful", zap.String("data-dir", cfg.DataDir))
+		}
+	}()
+
+	beConfig := backend.DefaultBackendConfig()
+	beConfig.Path = toBackendFileName(cfg.DataDir)
+	beConfig.Logger = cfg.Logger
+
+	be := backend.New(beConfig)
+	defer be.Close()
+
+	err = validateSchema(lg, be)
```

Comment:
This looks like a different implementation of `Verify` from main and `release-3.5` branch.
https://github.com/etcd-io/etcd/blob/main/server/verify/verify.go#L51
It's somewhat confusing for code readability across branches. And we are adding all this code to just check `confstate` and `term` buckets. 
Is this feature required? Looks like `nice-to-have`.

Reply:
This is to ensure we did not miss any part that let the 3.5 schema through. Adding this is more ensuring for supporting something that 3.4 does not support before. The feature is mentioned in https://docs.google.com/document/d/1siSo-tA_xLRo5li5wK8KqWsOiq34PIgtmuSDcyxxwsY/edit#heading=h.efr9asc472y. 

In theory, this schema verification should be in 3.5 and main as well once downgrade feature is complete. I don't see this as different implementation that confuses readability.

Reply:
1. I think original motivation for [Verify](https://github.com/etcd-io/etcd/commit/47b28b600adf210fe8e63c5484f63abea16d84fb) was ConsistencyIndex bug that was fixed later. IMO it's using env var as a guard because validateWal is expensive. There were no significant changes after original commit.
2. Your backport has completely different logic. Also removes `ExactIndex` in config.
3. At the moment there is no schema verification in 3.5/main Verify. But there is [schema.Validate](https://github.com/etcd-io/etcd/blob/main/server/etcdserver/bootstrap.go#L239) as part of NewServer logic. So, I'd add your validateSchema in server.NewServer after backend is initialized.

@ahrtr @serathius Please make a final call on this.

Reply:
Ideally, the best approach should be always implement in `main` firstly, and then backport to stable releases (3.5 and then 3.4). For this case, we should ideally
- Backport the [verification](https://github.com/etcd-io/etcd/blob/main/server/verify/verify.go) from `main` to 3.4 (already backported to 3.5); This can be a separate PR. Note that we are adding more and more changes in 3.4, this is definitely nice to have to add more protection & verification.
- Add the `validateSchema` in `main`, then backport to 3.5 and 3.4.  This can be discussed separately. (The `schema.Validate` is a different story, it was added in https://github.com/etcd-io/etcd/pull/13200 and never be backported to 3.5). 

For now, I don't want to block this PR as long as they can be addressed in followup PRs. Also https://github.com/etcd-io/etcd/pull/17330#issuecomment-1944436289